File: benchmarks/c/langbench.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

int main(int argc, char **argv) {
  if (argc < 3) {
    fprintf(stderr, "Usage: %s <filepath> <size>\n", argv[0]);
    return 1;
  }

  int size = atoi(argv[2]) / 2;
  if (size <= 0) {
    fprintf(stderr, "Error: Size must be a positive integer\n");
    return 1;
  }

  int *data = malloc(size * sizeof(int) * 2);

  FILE *file = fopen(argv[1], "r");
  if (!file) {
    fprintf(stderr, "Error: Could not open file %s\n", argv[1]);
    return 1;
  }
  fread(data, sizeof(int), size, file);
  fread(data + size, sizeof(int), size, file);
  fclose(file);

  double result = 0;
  for (size_t i = 0; i < size; i++) {
    result += sqrt(fabs(cos(data[i]) * sin(data[size + i])));
  }

  printf("%f\n", result);

  free(data);

  return 0;
}

File: benchmarks/nodejs/langbench.js
#!/usr/bin/env node
import fs from "node:fs/promises";
import { argv, exit } from "node:process";

if (argv.length < 4) {
  console.log(`Usage: ${argv[1]} <filepath> <size>`);
  exit(1);
}

const size = Math.floor(Number(argv[3]) / 2);
if (size <= 0) {
  console.log("Error: Size must be a positive integer");
  exit(1);
}

let left, right;
try {
  const file = await fs.open(argv[2], "r");
  const buffer = Buffer.alloc(size * 2 * 4); // 2 arrays, 4 bytes per int
  await file.read(buffer, 0, buffer.length, 0);
  await file.close();
  left = new Int32Array(buffer.buffer, 0, size);
  right = new Int32Array(buffer.buffer, size * 4, size);
} catch (err) {
  console.log(`Error: Could not open file ${argv[2]}`);
  exit(1);
}

let result = 0;
for (let i = 0; i < size; ++i) {
  result += Math.sqrt(Math.abs(Math.cos(left[i]) * Math.sin(right[i])));
}

console.log(result);
